internal/state/change_processor.go

```diff
@@ -60,23 +59,75 @@ type ChangeProcessorConfig struct {
 
 // ChangeProcessorImpl is an implementation of ChangeProcessor.
 type ChangeProcessorImpl struct {
-	store *store
-	cfg   ChangeProcessorConfig
+	cfg ChangeProcessorConfig
+
+	trackingUpdater *changeTrackingUpdater
+	clusterStore    graph.ClusterStore
+	updater         Updater
 
 	// changed is true if any changes that were captured require an update to nginx.
 	// It is true if the store changed, or if a Kubernetes resource (e.g.
 	// Service, EndpointSlice) that is related to a Gateway API resource (e.g. Gateway, HTTPRoute) changed.
 	// It is reset to false after Process is called.
-	changed bool
 
 	lock sync.Mutex
 }
 
 // NewChangeProcessorImpl creates a new ChangeProcessorImpl for the Gateway resource with the configured namespace name.
 func NewChangeProcessorImpl(cfg ChangeProcessorConfig) *ChangeProcessorImpl {
+	clusterStore := graph.ClusterStore{
+		GatewayClasses: make(map[types.NamespacedName]*v1beta1.GatewayClass),
+		Gateways:       make(map[types.NamespacedName]*v1beta1.Gateway),
+		HTTPRoutes:     make(map[types.NamespacedName]*v1beta1.HTTPRoute),
+		Services:       make(map[types.NamespacedName]*apiv1.Service),
+	}
+
+	trackingUpdater := newChangeTrackingUpdater(
+		cfg.GatewayClassName,
+		cfg.RelationshipCapturer,
+		[]changeTrackingUpdaterObjectTypeCfg{
+			{
+				gvk:                  mustGVKForObject(&v1beta1.GatewayClass{}),
+				store:                newObjectStoreImpl(clusterStore.GatewayClasses),
+				changingUpsertDelete: true,
+			},
+			{
+				gvk:                  mustGVKForObject(&v1beta1.Gateway{}),
+				store:                newObjectStoreImpl(clusterStore.Gateways),
+				changingUpsertDelete: true,
+			},
+			{
+				gvk:                  mustGVKForObject(&v1beta1.HTTPRoute{}),
+				store:                newObjectStoreImpl(clusterStore.HTTPRoutes),
+				changingUpsertDelete: true,
+			},
+			{
+				gvk:                  mustGVKForObject(&apiv1.Service{}),
+				store:                newObjectStoreImpl(clusterStore.Services),
+				changingUpsertDelete: false,
+			},
+			{
+				gvk:                  mustGVKForObject(&discoveryV1.EndpointSlice{}),
+				store:                nil,
+				changingUpsertDelete: false,
+			},
+		},
+	)
+
+	updater := newValidatingUpdater(
+		trackingUpdater,
+		[]upsertValidatorFunc{
+			func(obj client.Object) error {
+				// add re-run of webhook validation here
+				return nil
+			},
+		})
+
 	return &ChangeProcessorImpl{
-		store: newStore(),
-		cfg:   cfg,
+		cfg:             cfg,
+		trackingUpdater: trackingUpdater,
```

Comment:
I think storing a function here, like `getAndResetChanged()` from the first try, is better than storing the entire trackingUpdater. That way the get and reset are tied together which makes it harder to do the wrong thing, and we aren't exposing the other methods on the trackingUpdater to the changeProcessor. Those are all contained and called through the updater.